// coc-svelte/lib/src/tsplugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TsPlugin = void 0;
const coc_nvim_1 = require("coc.nvim");
class TsPlugin {
    constructor(context) {
        this.enabled = this.isEnabled();
        this.context = context;
        this.toggleTsPlugin(this.enabled);
        context.subscriptions.push(coc_nvim_1.workspace.onDidChangeConfiguration(() => {
            const enabled = this.isEnabled();
            if (enabled !== this.enabled) {
                this.enabled = enabled;
                this.toggleTsPlugin(this.enabled);
            }
        }));
    }
    isEnabled() {
        return coc_nvim_1.workspace.getConfiguration('svelte').get('enable-ts-plugin') ?? false;
    }
    async toggleTsPlugin(enable) {
        coc_nvim_1.workspace.getConfiguration('svelte').update('enable-ts-plugin', enable, true);
    }
    async askToEnable() {
        const shouldAsk = coc_nvim_1.workspace
            .getConfiguration('svelte')
            .get('ask-to-enable-ts-plugin');
        if (this.enabled || !shouldAsk) {
            return;
        }
        const answers = ['Ask again later', "Don't show this message again", 'Enable Plugin'];
        const response = await coc_nvim_1.window.showInformationMessage('The coc-svelte extension now contains a TypeScript plugin. ' +
            'Enabling it will provide intellisense for Svelte files from TS/JS files. ' +
            'Would you like to enable it? ' +
            'You can always enable/disable it later on through the extension settings.', ...answers);
        if (response === answers[2]) {
            coc_nvim_1.workspace.getConfiguration('svelte').update('enable-ts-plugin', true, true);
        }
        else if (response === answers[1]) {
            coc_nvim_1.workspace.getConfiguration('svelte').update('ask-to-enable-ts-plugin', false, true);
        }
    }
}
exports.TsPlugin = TsPlugin;
